File: rust/src/bin/ac04.rs
fn main() {
    let lines = std::io::stdin().lines();
    let cards: Vec<_> = lines.map(
	|line| {
	    let line = line.unwrap();
            let (_, contents) = line.as_str().split_once(':').unwrap();
            let (winning_numbers_str, numbers_str) =
                contents.split_once(" | ").unwrap();
	    let winning_numbers: std::collections::HashSet<i64> =
		winning_numbers_str.split(" ")
		.filter_map(|s| s.parse().ok()).collect();
	    let winning_numbers_count =
		numbers_str.split(" ").filter_map(|s| s.parse().ok())
		.filter(|number| winning_numbers.contains(number))
		.count();
	    winning_numbers_count
	}
    ).collect();
    let result_part1: usize = cards.iter().map(
	|&count| if count == 0 { 0 } else { 1 << (count - 1) }
    ).sum();
    let mut copies = Vec::new();
    let result_part2: usize = cards.iter().map(
	|&count| {
	    let copies_count = copies.pop().unwrap_or(1);
	    let previous_copies: Vec<_> =
		(0 .. count).map(|_| copies.pop().unwrap_or(1)).collect();
	    copies.extend(previous_copies.iter().rev().map(|v| v + copies_count));
	    copies_count
	}
    ).sum();
    println!("Part 1: {result_part1}");
    println!("Part 2: {result_part2}");
}

File: haskell/ac04.hs
import Data.Char
import Data.List
import Data.Bits
import qualified Data.Set as Set

cut p list =
  let (hd, _:tl) = break p list
  in (hd, tl)

split p list =
  let (hd, tl) = break p list
  in case tl of
    [] -> [hd]
    _:tl -> hd:split p tl

split_numbers = map read . filter (/= []) . split (== ' ')

parseLine line =
  let (_, contents) = cut (== ':') line
      (winning, cards) = cut (== '|') (dropWhile isSpace contents)
      winning_numbers :: [Int] = split_numbers (dropWhileEnd isSpace winning)
      card_numbers :: [Int] = split_numbers (dropWhile isSpace cards)
      winning_set = Set.fromList winning_numbers
  in length (filter (flip Set.member winning_set) card_numbers)

scoreGame :: Int -> Int
scoreGame winningCount =
  if winningCount == 0 then
    0
  else
    shiftL 1 (winningCount - 1)

countCards total copies winningCounts =
  case (winningCounts, copies) of
    ([], _) -> total
    (hd:tl, copies:copies_tail) ->
      let list = map (+ copies) (take hd copies_tail) ++ drop hd copies_tail
      in countCards (total + copies) list tl

main = do
  input <- getContents
  let input_lines = lines input
  let winningCounts = map parseLine input_lines
  let result_part1 = sum (map scoreGame winningCounts)
  let result_part2 = countCards 0 (repeat 1) winningCounts
  putStrLn ("Part 1: " ++ show result_part1)
  putStrLn ("Part 2: " ++ show result_part2)
